### PR TITLE
Feature/rjm task exclusion

### PR DESCRIPTION
**Pull request by robin-miller-ow**
Hi Luke,

Please enjoy this PR.

Thanks,

Robin

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -762,6 +762,19 @@ def init_on_load(self):
         """ Initialize the attributes that aren't stored in the DB. """
         self.test_mode = False  # can be changed when calling 'run'
 
+    def state_for_dependents(self):
+        """
+        Helper function used to wire in new EXCLUDED state. For identifying
```
Comment (lukem-ow) [inline airflow/models.py]: This comment will probably be in the code a while so I am not sure 'new' will always be relevant.

File: airflow/models.py [hunk 2]
```diff
@@ -762,6 +762,19 @@ def init_on_load(self):
...
+        whether task dependencies are met the EXCLUDED state should be treated
```
Comment (lukem-ow) [inline airflow/models.py]: This sentence doesn't seem to read correctly.

File: airflow/models.py [hunk 3]
```diff
@@ -762,6 +762,19 @@ def init_on_load(self):
...
+        as SUCCESS. This function allows us to encompass this logic in one
+        place.
+        :return: the effective state of the task instance.
+        """
+        if self.state == State.EXCLUDED:
+            return State.SUCCESS
```
Comment (lukem-ow) [inline airflow/models.py]: This seems like an odd thing to do. Should we have logic elsewhere that handles EXCLUDED separately to SUCCESS

File: airflow/models.py [hunk 4]
```diff
@@ -3975,3 +3988,247 @@ class ImportError(Base):
...
+            task_id,
+            exclusion_type,
+            exclusion_start_date,
+            exclusion_end_date,
+            created_by,
+            session=None):
+        """
+        Add a task exclusion to prevent a task running under certain
+        circumstances.
+        :param dag_id: The dag_id of the DAG containing the task to exclude
```
Comment (lukem-ow) [inline airflow/models.py]: no information on the cls parameter.

File: airflow/www/templates/airflow/dag.html [hunk 5]
```diff
@@ -211,6 +211,20 @@ <h4 class="modal-title" id="myModalLabel">
               Recursive
             </button>
           </span>
+          <hr/>
+          <button id="btn_exclude" type="button" class="btn btn-primary">
+            Exclude From
+          </button>
+          <span class="btn-group">
+            <button id="btn_exclude_from_current"
```
Comment (lukem-ow) [inline airflow/www/templates/airflow/dag.html]: I think we can remove this button.

File: airflow/models.py [hunk 6]
```diff
@@ -3975,3 +3988,247 @@ class ImportError(Base):
...
+          more detail.
+        :param exclusion_start_date: The execution_date that the exclusion
+         starts on. This will be ignored if the exclusion_type is INDEFINITE.
+        :param exclusion_end_date: The execution_date that the exclusion ends
+         on. This will be ignored if the exclusion_type is INDEFINITE or
+         SINGLE_DATE.
+        :return: None.
+        """
+
+        session.expunge_all()
```
Comment (lukem-ow) [inline airflow/models.py]: Is it standard practice to remove all objects from the session?

File: airflow/models.py (hunk 1 above)
Comment (robin-miller-ow) [inline airflow/models.py]: Agreed. Will fix.

File: airflow/models.py (hunk 2 above)
Comment (robin-miller-ow) [inline airflow/models.py]: As discussed, comma needed. Will fix

File: airflow/models.py (hunk 3 above)
Comment (robin-miller-ow) [inline airflow/models.py]: I am making the assumption that for flow control purposes exclusion is equivalent to success. Do you believe that should not be the case?

File: airflow/models.py (hunk 4 above)
Comment (robin-miller-ow) [inline airflow/models.py]: It's a python thing, you don't pass it in manually. I'm failing to remember whether the automatically produced template included it or not. I shall check. If not, I would posit that documentation is not expected. If it is, I'll add some.

File: airflow/www/templates/airflow/dag.html (hunk 5 above)
Comment (robin-miller-ow) [inline airflow/www/templates/airflow/dag.html]: Agreed, it'll be easy to add back in when we do want it.

File: airflow/models.py (hunk 4 above)
Comment (robin-miller-ow) [inline airflow/models.py]: (It doesn't add it automatically)

File: airflow/models.py (hunk 3 above)
Comment (lukem-ow) [inline airflow/models.py]: I believe it should but it just seems wrong to return a state of success if the state is actually excluded.

File: airflow/models.py (hunk 4 above)
Comment (lukem-ow) [inline airflow/models.py]: It is being used below and I have no idea what it is.

Comment (robin-miller-ow) [inline airflow/models.py]: As discussed, this is a python language feature and should not need documenting within the code.

File: airflow/models.py (hunk 3 above)
Comment (robin-miller-ow) [inline airflow/models.py]: As discussed, this is an attempt to move this logic into a single location instead of everywhere and is acceptable for now.

**Pull request merged**